Complete with  test _unittest_raft_leader_changes()

Parsifal22 opened the pull request:
In node.py: expanded the log information for some values ​​to make it easier to analyze the code

In raft_log_replication.py: start to write test that log replication happens correctly if leadership changes

In raft_node.py: add a couple of lines of code to close node after each test, so that after running pytest it does not crash

new issue: after the leader changes, the new leader is not aware of the latest index of the neighboring nodes and begins to rewrite them, even though his nodes completely coincide with the neighboring node.

Parsifal22 commented:
Test case: 
There are 3 nodes in the cluster: 41, 42, 43. At first, the leader of node 41 receives 3 elements and then distributes them to followers. After that, node 42 becomes the leader and log files should be saved at all nodes.

Problem:
Each leader has its own variable self._next_index it contains the values ​​of the current index for each follower. If the index is similar to the current user index, then it sends the heartbeat function, if not, then it starts sending its logs to the follower until there is an equal number:
![image](https://github.com/OpenCyphal-Garage/cyraft/assets/73550345/bccc1fce-df76-4b6b-addb-916fabf913c8)

How it looks in the logs:
![image](https://github.com/OpenCyphal-Garage/cyraft/assets/73550345/5c1b75d2-d8d7-43fb-95de-5bca2f79ad17)

When the leader changes to 42 nodes, their variable self._next_index with all the info about other logs is empty and it starts rewriting them again.
How it looks in the logs:
![image](https://github.com/OpenCyphal-Garage/cyraft/assets/73550345/a4cc3a6d-0535-4c65-a186-d6d7ddd1010d)

Since this prevented him from sending a heartbeat, other followers start putting forward their candidacy until 41 becomes the leader again, since his logs are updated and he can safely send a heartbeat signal.

maksimdrachov commented on cyraft/node.py:
```diff
@@ -442,7 +454,9 @@ def _append_entries_processing(
 
         # Update current_term (if follower) (leaders will update their own term on timeout)
         if self._state == RaftState.FOLLOWER:
+            self._change_state(RaftState.FOLLOWER)  # this will reset the election timeout as well
```
I don't think using `_change_state()` in this way is correct.

In regard to reseting the election timeout, there are three cases that we have to take into account:

1. RPC request or response contains `term T > currentTerm` (§5.1, page 4)

    -> convert to Follower (this will also reset the election timeout)

2. Regular Heartbeat (= empty AppendEntries)

    -> reset the election timeout (no `_change_state()`)

3. Regular AppendEntries

    -> rest the election timeout (no `_change_state()`)

I would suggest to use `self._reset_election_timeout()` directly.

maksimdrachov commented on cyraft/node.py:
```diff
@@ -88,6 +88,8 @@ def __init__(self) -> None:
 
         ## Volatile state on leaders
         self._next_index: typing.List[int] = []  # index of the next log entry to send to that server
+
+
         # self._match_index: typing.List[int] = []  # index of highest log entry known to be replicated on server
```
Actually now that I think of it: we're always using Black formatter for all of the code.

In order to avoid these formatting type of comments/discussions.

See pycyphal:

<img width="637" alt="image" src="https://github.com/OpenCyphal-Garage/cyraft/assets/39975120/9a717f38-d0cc-421e-a3ce-4ce0a4882645">

(This is also something that should be part of CI, code that is not formatted correctly should fail)

maksimdrachov commented on tests/raft_log_replication.py:
```diff
@@ -593,7 +593,419 @@ async def _unittest_raft_log_replication() -> None:
+    raft_node_1._change_state(RaftState.FOLLOWER)
+    raft_node_2._change_state(RaftState.FOLLOWER)
+    raft_node_3._change_state(RaftState.FOLLOWER)
+    raft_node_1._voted_for = None
```
Why is this necessary? (Should't it be able to know that it hasn't voted for anyone just be being made a Follower?)

maksimdrachov commented on tests/raft_log_replication.py:
```diff
@@ -593,7 +593,419 @@ async def _unittest_raft_log_replication() -> None:
+
+    await asyncio.sleep(TERM_TIMEOUT + 1)
+
+    assert len(raft_node_2._log) == 1 + 4
+    assert raft_node_2._log[0].term == 0
+    assert raft_node_2._log[0].entry.value == 0
+    assert raft_node_2._log[1].term == 4
+    assert raft_node_2._log[1].entry.name.value.tobytes().decode("utf-8") == "top_1"
+    assert raft_node_2._log[1].entry.value == 7
+    assert raft_node_2._log[2].term == 5
+    assert raft_node_2._log[2].entry.name.value.tobytes().decode("utf-8") == "top_2"
+    assert raft_node_2._log[2].entry.value == 8
+    assert raft_node_2._log[3].term == 6
+    assert raft_node_2._log[3].entry.name.value.tobytes().decode("utf-8") == "top_3"
+    assert raft_node_2._log[3].entry.value == 9
+    assert raft_node_2._log[4].term == 7
+    assert raft_node_2._log[4].entry.name.value.tobytes().decode("utf-8") == "top_4"
+    assert raft_node_2._log[4].entry.value == 13
+    assert raft_node_2._commit_index == 4
```
From the comment above the test I was under the impression it would result in the following:

```
   Step 3: Replace Log Entry 3 with a New Entry fom LEADER Node 42
   ____________
  | 3          |     Log index
  | 7          |     Log term
  | top_3 <= 10|     Name <= value
  |____________|

  Result:
   ____________ ____________ ____________ _____________
  | 0          | 1          | 2          | 3           |     Log index
  | 0          | 4          | 5          | 7           |     Log term
  | empty <= 0 | top_1 <= 7 | top_2 <= 8 | top_3 <= 10 |     Name <= value
  |____________|____________|____________|_____________|
```

maksimdrachov commented on tests/raft_log_replication.py:
```diff
@@ -593,7 +593,419 @@ async def _unittest_raft_log_replication() -> None:
+    assert raft_node_3._log[4].entry.value == 13
+    assert raft_node_3._commit_index == 4
+
+    _logger.info("================== TEST 3: Replace log entries 1 and 4 with new entries from another LEADER ==================")
```
Make sure that the comment which documents the test and the actual test do align, I was promised the following in the comment above the unit test:

```
Step 3: Replace Log Entry 3 with a New Entry fom LEADER Node 42
```

maksimdrachov commented on tests/raft_node.py:
```diff
@@ -155,123 +157,6 @@ async def _unittest_raft_node_election_timeout() -> None:
     await asyncio.sleep(1)  # give some time for the node to close
 
 
-async def _unittest_raft_node_heartbeat() -> None:
```
Why did you remove this test?

Parsifal22 commented:
Fixed these small issues:
Removed unnecessary spaces between lines of code
Replaced the function self._change_state(RaftState.FOLLOWER) with the function self._reset_election_timeout() in cyraft/node.py
Reimplemented the _unittest_raft_node heartbeat() test to tests/raft_node.py

maksimdrachov commented:
These commits related to CI shouldn't be here :)

maksimdrachov commented on tests/raft_log_replication.py:
```diff
@@ -593,7 +593,422 @@ async def _unittest_raft_log_replication() -> None:
+    raft_node_3.close()
+    await asyncio.sleep(1)
+
+# =======================================================================================
+# ========== Test that log replication happens correctly if leadership changes ==========
+# =======================================================================================
+
```
Move this into the comment block below, it makes little sense to have these in 2 separate blocks.

Parsifal22 replied:
Merged two comment blocks into one

maksimdrachov reviewed (changes requested):
Almost there :)

Parsifal22 commented:
> Test case: There are 3 nodes in the cluster: 41, 42, 43. At first, the leader of node 41 receives 3 elements and then distributes them to followers. After that, node 42 becomes the leader and log files should be saved at all nodes.
> 
> Problem: Each leader has its own variable self._next_index it contains the values ​​of the current index for each follower. If the index is similar to the current user index, then it sends the heartbeat function, if not, then it starts sending its logs to the follower until there is an equal number: ![image](https://private-user-images.githubusercontent.com/73550345/343771995-bccc1fce-df76-4b6b-addb-916fabf913c8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA0NDQ4NTYsIm5iZiI6MTcyMDQ0NDU1NiwicGF0aCI6Ii83MzU1MDM0NS8zNDM3NzE5OTUtYmNjYzFmY2UtZGY3Ni00YjZiLWFkZGItOTE2ZmFiZjkxM2M4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzA4VDEzMTU1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVkODJmNGUzMDc1N2M2MGFlZjE1ZjZlMzYwZTBlNjhiZjZmNmIxOWE1ZWZlYTc0Mjg4MDY2MDJmMTI4ZGVhODgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.3cPfSf9cCqBLIxLsr7TM4PIN-ZLqyWDiwWzLN82Cx74)
> 
> How it looks in the logs: ![image](https://private-user-images.githubusercontent.com/73550345/343774298-5c1b75d2-d8d7-43fb-95de-5bca2f79ad17.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA0NDQ4NTYsIm5iZiI6MTcyMDQ0NDU1NiwicGF0aCI6Ii83MzU1MDM0NS8zNDM3NzQyOTgtNWMxYjc1ZDItZDhkNy00M2ZiLTk1ZGUtNWJjYTJmNzlhZDE3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzA4VDEzMTU1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZhOTUwYjYwODVhZWVlMDRlZDEwMDk1MzhkM2Y1NDNmMTczZmI2ZDkwN2VkYzQ3OWIxMjk4Yjg0MmQ3OThjODUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.0DlA3wDFt_SO8jPxKK2VQ4g4yHiFsV_5JKZKg8w23SM)
> 
> When the leader changes to 42 nodes, their variable self._next_index with all the info about other logs is empty and it starts rewriting them again. How it looks in the logs: ![image](https://private-user-images.githubusercontent.com/73550345/343778609-a4cc3a6d-0535-4c65-a186-d6d7ddd1010d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA0NDQ4NTYsIm5iZiI6MTcyMDQ0NDU1NiwicGF0aCI6Ii83MzU1MDM0NS8zNDM3Nzg2MDktYTRjYzNhNmQtMDUzNS00YzY1LWExODYtZDZkN2RkZDEwMTBkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzA4VDEzMTU1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY2ZjY0N2ExMjE2ZWNmMDBhNTU5MjFiNmZhMTMxMjUxYTY5Mjk2OWU0ZWZhMmVmMWVmMDUxOTk4ZGMxNGNmNDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Q8doMWR5fBOipVZy6fd0gb-Cqm3Qjkw2gX7F2aqaiGo)
> 
> Since this prevented him from sending a heartbeat, other followers start putting forward their candidacy until 41 becomes the leader again, since his logs are updated and he can safely send a heartbeat signal.

New errors related to log replication have appeared. The error appears in the _unittest_raft_leader_changes() test in raft_log_replication.py. Case description: We have 4 nodes (41, 42, 43, 44). At the beginning, the leader node is 41, which receives and replicates 3 logs. After that, it disappears from the cluster:
![image](https://github.com/OpenCyphal-Garage/cyraft/assets/73550345/feac7bf4-2220-46b7-9649-bbda2129929d)

the new leader after this becomes node 42, which begins to rewrite data about all logs for all nodes as described above, except for node 41, since it is not yet in our cluster and for it the value of the next index is 1:
![image](https://github.com/OpenCyphal-Garage/cyraft/assets/73550345/fe71c631-bfdb-4592-bbbb-7c4fe314463d)

Then we add a new log from node 42 and rewrite the third log. When the third log is rewritten, it automatically removes this value from the list and the following ones too (in our case, only 4) as shown in this code:
![image](https://github.com/OpenCyphal-Garage/cyraft/assets/73550345/d2a8299f-4e0f-4d58-b2f7-5db155076974)

the self._next_index array exists so that each node knows about other nodes what their next log index is. Since node 41 has a next index of 1, then its next index is -1, which creates a conflict in AppendEntries, since the prev_log_index value we send is self._next_index[remote_node_index] - 1, which gives a value of -2, and the data type of prev_log_index is uint64. In the end, we get this error:
![image](https://github.com/OpenCyphal-Garage/cyraft/assets/73550345/71d6142f-4200-4d0b-a324-c65988a97059)

Parsifal22 commented on tests/raft_log_replication.py:
```diff
@@ -593,7 +593,441 @@ async def _unittest_raft_log_replication() -> None:
+
+    raft_node_1.close() # Simulation of the disappearance of a leader from a cluster
+
+    await asyncio.sleep(ELECTION_TIMEOUT + 6*TERM_TIMEOUT + 0.3)
```
Six terms are needed for complete replication of logs from the leader to the followers. The value 0.3 was chosen because before with the value 0.1 the test worked every other time. Now, [in the new commit](https://github.com/OpenCyphal-Garage/cyraft/pull/11/commits/42200f3fbf02ea66f1209ad04ee87873798a38cb#:~:text=await%20asyncio.sleep(ELECTION_TIMEOUT%20%2B%206*TERM_TIMEOUT%20%2B%200.1)%20%23%20Six%20terms%20are%20needed%20for%20complete%20replication%20of%20logs%20from%20the%20leader%20to%20the%20followers.), the work is stable even with the value 0.1 (I ran the test 4 times with the value +0.1 and never gave an error).

Parsifal22 commented on tests/raft_log_replication.py:
```diff
@@ -593,7 +593,441 @@ async def _unittest_raft_log_replication() -> None:
+    # New LEADER => raft_node_2
+    raft_node_2.election_timeout = ELECTION_TIMEOUT
+    raft_node_3.election_timeout = ELECTION_TIMEOUT + 2
+    raft_node_4.election_timeout = ELECTION_TIMEOUT + 2
```
I corrected the test [conditions](https://github.com/OpenCyphal-Garage/cyraft/pull/11/commits/2093d422adfd05814e9cc0daa470ca8c26429d86#:~:text=raft_node_1.election_timeout,2.1). But node 4 increased by +2.1, because node 2 launched the _serve_request_vote_impl method and node 4 launched the _start_election method in parallel, which is why the `self._voted_for` value was assigned the value of node 4 instead of None, which is why the request did not pass [the condition](https://github.com/OpenCyphal-Garage/cyraft/pull/11/commits/024fd95896d70b8fa74c7c8a52d69c40c4a24b76#:~:text=vote_granted%20%3D%20(self._voted_for%20is%20None%20or%20self._voted_for%20%3D%3D%20client_node_id)%20and%20self._log%5Brequest.last_log_index%5D.term%20%3D%3D%20request.last_log_term) and node 2 lost the election.

maksimdrachov commented on tests/raft_log_replication.py:
```diff
@@ -593,7 +593,441 @@ async def _unittest_raft_log_replication() -> None:
+    _logger.info("================== TEST 2: Leadership Change to Node 42 and Add New Entry  ==================")
+
+    # New LEADER => raft_node_2
+    raft_node_1.election_timeout = ELECTION_TIMEOUT
+    raft_node_2.election_timeout = ELECTION_TIMEOUT + 1
+    raft_node_3.election_timeout = ELECTION_TIMEOUT + 2
+    raft_node_4.election_timeout = ELECTION_TIMEOUT + 2.1
+    await asyncio.sleep(TERM_TIMEOUT)
```
I meant that you define the election timeouts for all the nodes only one time at this test and after that don't touch those anymore.

maksimdrachov commented on tests/raft_log_replication.py:
```diff
@@ -593,7 +593,441 @@ async def _unittest_raft_log_replication() -> None:
+
+    raft_node_1.close() # Simulation of the disappearance of a leader from a cluster
+
+    await asyncio.sleep(ELECTION_TIMEOUT + 8*TERM_TIMEOUT + 0.1) # Six terms are needed for complete replication of logs from the leader to the followers.
```
Ok, I think I understand your issue here.

Currently when a node becomes a new leader it assumes all the other nodes are behind/have empty logs.

If you think about it, this only makes sense at the start, after that the nodes will always be in various degrees of up-to-dateness.

This means that when a node becomes Leader, the best strategy is to assume all other nodes _are_ up-to-date, until proven otherwise. After sending the first Heartbeat/AppendEntryRequest, you will now whether that is true:

<img width="443" alt="Screenshot 2024-07-10 at 16 42 51" src="https://github.com/OpenCyphal-Garage/cyraft/assets/39975120/28053e02-7f8a-4177-b4b8-4108abc5682f">

Then it should start to send AppendEntriesRequest in decreasing order, I think this part is already implemented:

<img width="775" alt="image" src="https://github.com/OpenCyphal-Garage/cyraft/assets/39975120/a70a59c4-0db5-4e1f-a61d-570747891bd2">

Now I'm also realising we're only testing the case where all the other nodes are indeed up-to-date. So therefore it would make sense to, once `node_2` has been elected as leader, to clear `node_1`'s log and re-start it, this node should then also be brought up to date by `node_2`.

Parsifal22 commented:
In this [commit](https://github.com/OpenCyphal-Garage/cyraft/pull/11/commits/73c4920ce852609999f1cc00bdd4b9751283c715) i updated the term

maksimdrachov commented on tests/raft_log_replication.py:
```diff
@@ -196,191 +183,44 @@ async def _unittest_raft_log_replication() -> None:
-    assert raft_node_3._log[3].entry.value == 12
-
-    _logger.info("================== TEST 4: Add an already existing log entry ==================")
+    _logger.info("================== TEST 2: Add an already existing log entry ==================")
```
So you just remove a test if it's not passing, or what?

![image](https://github.com/user-attachments/assets/31be285c-48bb-434a-8d84-e44f75148df2)

Parsifal22 replied:
These tests are no longer relevant -> [answer](https://github.com/OpenCyphal-Garage/cyraft/pull/11#discussion_r1679520402)

maksimdrachov commented on tests/raft_log_replication.py:
```diff
@@ -34,28 +34,15 @@
-  |____________|____________|
 
- Step 4: Add an already existing log entry
+ Step 2: Add an already existing log entry
```
It _used_ to exist, since you removed the block where it got added, it doesn't anymore.

Parsifal22 replied:
Solved in branch fix-term commit 8d01fa6

maksimdrachov commented on tests/raft_log_replication.py:
```diff
@@ -626,20 +527,6 @@ async def _unittest_raft_log_replication() -> None:
   | 0          | 4          | 5          | 6          | 7           |     Log term
```
Interesting, how the terms in the comment are all different, and in the test they're all the same (1).

Parsifal22 replied:
Solved in branch fix-term commit 8d01fa6

Parsifal22 commented on cyraft/node.py:
```diff
@@ -231,7 +231,7 @@ async def _serve_request_vote_impl(
 
             if vote_granted:
-                self._change_state(RaftState.FOLLOWER) # Avoiding race condition when Candidate. This is necessary to avoid excessive elections
+                #self._change_state(RaftState.FOLLOWER) # Avoiding race condition when Candidate. This is necessary to avoid excessive elections
```
Resolved this issue on fix-term branch in commit 304d33c

Parsifal22 commented on cyraft/node.py:
```diff
@@ -395,11 +397,14 @@ def _append_entries_processing(
+        new_index = request.prev_log_index + 1
+        _logger.info(c["append_entries"] + "Node ID: %d -- new_index: %d" + c["end_color"], self._node.id, new_index)
 
+        #a leader never overwrites or deletes entries in its log; it only appends new entries
```
[The paper](https://raft.github.io/raft.pdf) states that the leader can only add a new entity, but not rewrite or delete:
![Figure 3](https://github.com/user-attachments/assets/448979c8-0034-4095-9f24-9d7264826353)

Plus it is written the same paper that entity can only be rewritten if the index is the same and the term is different, but the term only changes after the elections, when a new leader comes.
![image](https://github.com/user-attachments/assets/67a3dd47-4d5b-45bc-a04c-293251e045e3)

maksimdrachov replied:
> Plus it is written the same paper that entity can only be rewritten if the index is the same and the term is different, but the term only changes after the elections, when a new leader comes.

But then why mention this at all in the implementation?

![image](https://github.com/user-attachments/assets/830eabcb-944d-438d-b90e-0be1f636e17c)

I'm pretty sure that this rule is necessary in some bordercase where a node (or a group of nodes) becomes disconnected from the larger network, they proceed to choose a leader, updates its log to something and then reconnects to the larger network (something like that). In that case it is probably necessary that the leader is able to overwrite the log of its followers.

Parsifal22 replied:
Yes. But the implementation of this function was incorrect. Now, every time entities are added to the log, it starts a cycle that checks each element in the log. This is not efficient, because if we have 20 nodes with 10,000 logs and we add another entity, then to put it mildly, it is not very fast, especially for RTOS

maksimdrachov replied:
inefficient != incorrect 

Log comparison could indeed be done more efficiently, so if you want to optimize it, go ahead.

Parsifal22 commented on cyraft/node.py:
```diff
@@ -661,11 +696,19 @@ def _reset_election_timeout(self) -> None:
-        self._election_timer = loop.call_at(
-            self._next_election_timeout, lambda: asyncio.create_task(self._on_election_timeout())
+        self._election_timer = loop.call_later(
+            self._election_timeout, lambda: asyncio.create_task(self._on_election_timeout())
```
Look what I wrote in the comments:
**When on_election_timeout is called, the state is changed to candidate and a new election is started by calling start_election function. In the start_election function, the election timeout is reset using self._next_election_timeout = loop.time() self._election_timeout.  If loop.time() returns a value greater than or equal to self._next_election_timeout, then the election timeout will be set to a time in the past, causing the timeout to trigger immediately. To fix this issue, added a small offset to the current time when resetting the election timeout**
From the next commit I will remove this expression `self._next_election_timeout = loop.time() + self._election_timeout + 0.01` because the `loop.call_later` function is more suitable than `loop.call_at`. It does not need to calculate the next timeout as for `loop.call_at`.

Parsifal22 commented on cyraft/node.py:
```diff
@@ -661,11 +696,19 @@ def _reset_election_timeout(self) -> None:
+        """
+        When on_election_timeout is called, the state is changed to candidate and a new election is started by calling start_election function. 
+        In the start_election function, the election timeout is reset using self._next_election_timeout = loop.time() + self._election_timeout. 
+        If loop.time() returns a value greater than or equal to self._next_election_timeout, then the election timeout will be set to a time in the past, 
```
Maybe you are right. But when *candidate 42* sent a request with a higher term to *node 43*, then it reset the timer and instantly ran the `_on_election_timeout()` method, which started the election for *candidate 43*. The instant call [in this line](https://github.com/OpenCyphal-Garage/cyraft/pull/11/files#:~:text=self._election_timer%20%3D,self._on_election_timeout())) can happen because of the characteristics of the `loop.call_at` method, since if we specify a later or negative time, it runs `self._on_term_timeout()` immediately. That's why I decided to use `loop.call_later`, since it configures the period after which the method is run, and not just configures the time to run the method.

maksimdrachov replied:
> call [in this line](https://github.com/OpenCyphal-Garage/cyraft/pull/11/files#:~:text=self._election_timer%20%3D,self._on_election_timeout()) can happen

Link doesn't work: https://github.com/OpenCyphal-Garage/cyraft/pull/11/files#:~:text=self._election_timer%20%3D,self._on_election_timeout()

It should look something like this: https://github.com/Parsifal22/cyraft/blob/67a48aa8fb28c0c0b070733da4da338e747eb738/cyraft/node.py#L702

<img width="575" alt="image" src="https://github.com/user-attachments/assets/f6ed72e8-6883-4b81-8fc9-bd6949eecc9a">

maksimdrachov replied:
If instead of using `call_at`, you'd like to use `call_later`, that's okay, but functionally it should do exactly the same:

<img width="832" alt="image" src="https://github.com/user-attachments/assets/f8d7fcb2-8cbf-423e-8410-4729771c5c53">

https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.call_at

As far as I can tell:

```py
def _reset_election_timeout(self) -> None:
        """
        If a follower receives a heartbeat from the leader, it should reset its election timeout.
        """
        assert self._state == RaftState.FOLLOWER, "Only followers should reset the election timeout"
        _logger.info(c["raft_logic"] + "Node ID: %d -- Resetting election timeout" + c["end_color"], self._node.id)
        loop = asyncio.get_event_loop()
        self._next_election_timeout = loop.time() + self._election_timeout
        if hasattr(self, "_election_timer"):
            self._election_timer.cancel()
        self._election_timer = loop.call_at(
            self._next_election_timeout, lambda: asyncio.create_task(self._on_election_timeout())
        )
```

Is the same as this:

```py
def _reset_election_timeout(self) -> None:
        """
        If a follower receives a heartbeat from the leader, it should reset its election timeout.
        """
        assert self._state == RaftState.FOLLOWER, "Only followers should reset the election timeout"
        _logger.info(c["raft_logic"] + "Node ID: %d -- Resetting election timeout" + c["end_color"], self._node.id)
        loop = asyncio.get_event_loop()
        if hasattr(self, "_election_timer"):
            self._election_timer.cancel()
        self._election_timer = loop.call_later(
            self._election_timeout, lambda: asyncio.create_task(self._on_election_timeout())
```

So you can keep this.

Parsifal22 commented on cyraft/node.py:
```diff
@@ -414,6 +419,8 @@ def _append_entries_processing(
                 self._commit_index = log_index
                 break
+        """
+
```
Solved in branch fix-term commit 8d01fa6

Parsifal22 commented on tests/raft_leader_election.py:
```diff
@@ -396,17 +394,17 @@ async def _unittest_raft_fsm_4():
     assert raft_node_2._prev_state == RaftState.CANDIDATE
     assert raft_node_2._state == RaftState.LEADER
-    assert raft_node_2._term == 12, "+ 1 due to election timeout, + 10 due to term timeout"
+    assert raft_node_2._term == 2, "+ 1 due to election timeout, + 10 due to term timeout"
```
Solved in branch fix-term commit 8d01fa6

maksimdrachov commented on cyraft/node.py:
```diff
@@ -685,12 +753,22 @@ async def _on_term_timeout(self) -> None:
             self._term,
         )
         assert self._state == RaftState.LEADER, "Only leaders have a term timeout"
-        self._term += 1
```
This change is already part of main, please make sure you have rebased correctly!

maksimdrachov commented on cyraft/election_watchdog.py:
```diff
@@ -29,7 +31,9 @@ async def run(self) -> None:
         _logger.debug("Starting election watchdog")
         loop = asyncio.get_event_loop()
         self._next_election_timeout = loop.time() + self.election_timeout
-        self._election_timer = loop.call_at(self._next_election_timeout, self.on_election_timeout)
+        self._election_timer = loop.call_at(
+            self._next_election_timeout, self.on_election_timeout
+        )
```
Ok, please use the following settings in VSCode:

```
{
    "editor.defaultFormatter": "ms-python.black-formatter",
    "black-formatter.args": [
        "--line-length=120"
    ],
    "editor.formatOnSave": true
}
```

I might have forgotten to mention we use a line length of 120.

https://wiki.zubax.com/en/public/development/Zubax-Python-coding-conventions

Parsifal22 replied:
[Fixed](https://github.com/OpenCyphal-Garage/cyraft/pull/11/commits/e9cd8165d7ba66c545d75d3037f5b044045cc01d)

maksimdrachov commented on cyraft/election_watchdog.py:
```diff
@@ -43,11 +47,15 @@ async def _unittest_election_watchdog() -> None:
     watchdog_node = ElectionWatchdog()
     asyncio.create_task(watchdog_node.run())
 
-    await asyncio.sleep(ELECTION_TIMEOUT - 1)  # Election timeout should not be reached here
+    await asyncio.sleep(
+        ELECTION_TIMEOUT - 1
+    )  # Election timeout should not be reached here
 
     watchdog_node.reset_election_timeout()  # Reset election timeout
 
-    await asyncio.sleep(1)  # Election timeout should be reached here, but it should be reset above
+    await asyncio.sleep(
+        1
+    )  # Election timeout should be reached here, but it should be reset above
```
This will be re-formatted after adjusting line length.

Parsifal22 replied:
[Fixed](https://github.com/OpenCyphal-Garage/cyraft/pull/11/commits/e9cd8165d7ba66c545d75d3037f5b044045cc01d)

maksimdrachov commented on tests/raft_node.py:
```diff
@@ -156,124 +167,6 @@ async def _unittest_raft_node_election_timeout() -> None:
     await asyncio.sleep(1)  # give some time for the node to close
 
 
-async def _unittest_raft_node_heartbeat() -> None:
```
Why did you remove this test?

Parsifal22 replied:
I don't know why it was removed but in [this commit](https://github.com/OpenCyphal-Garage/cyraft/pull/11/commits/6073cca7fdab641e62ff235075cafbabc6d40956) I added it back

Parsifal22 replied:
[Fixed](https://github.com/OpenCyphal-Garage/cyraft/pull/11/commits/e9cd8165d7ba66c545d75d3037f5b044045cc01d)

maksimdrachov reviewed (commented):
Ok, I think it looks good, can you do the following:

- [ ] Formatting is re-stored to 120 (sorry, I forgot to mention this)
- [ ] Heartbeat test is put back? (I don't see a good reason why it should be removed?)

After that I will review once more, and I think it should be finished.

maksimdrachov commented on tests/raft_node.py:
```diff
@@ -118,15 +124,20 @@ async def _unittest_raft_node_term_timeout() -> None:
     raft_node._change_state(RaftState.LEADER)  # only leader can increase term
 
     await asyncio.sleep(TERM_TIMEOUT)  # + 0.1 to make sure the timer has been reset
-    assert raft_node._term == 0 # 0 because we have manually set the node to leader (instead of waiting for election)
+    assert (
+        raft_node._term == 0
+    )  # 0 because we have manually set the node to leader (instead of waiting for election)
```
format

maksimdrachov commented on tests/raft_node.py:
```diff
@@ -118,15 +124,20 @@ async def _unittest_raft_node_term_timeout() -> None:
 
     raft_node._change_state(RaftState.FOLLOWER)  # follower should not increase term
     await asyncio.sleep(TERM_TIMEOUT)
     assert raft_node._term == 0
+    assert raft_node._term == 0
```
Why do we require the same statement twice?

maksimdrachov commented on tests/raft_node.py:
```diff
@@ -169,7 +255,9 @@ async def _unittest_raft_node_heartbeat() -> None:
     raft_node._voted_for = 42
 
     asyncio.create_task(raft_node.run())
-    await asyncio.sleep(ELECTION_TIMEOUT * 0.90)  # sleep until right before election timeout
+    await asyncio.sleep(
+        ELECTION_TIMEOUT * 0.90
+    )  # sleep until right before election timeout
```
format

maksimdrachov commented on tests/raft_node.py:
```diff
@@ -196,8 +284,12 @@ async def _unittest_raft_node_heartbeat() -> None:
 
     # send heartbeat again
     # (this time leader has a higher term, we want to make sure that the follower's term is updated)
-    await asyncio.sleep(ELECTION_TIMEOUT * 0.90)  # sleep until right before election timeout
-    terms_passed = raft_node._term + 5  # leader's term is higher than the follower's term
+    await asyncio.sleep(
+        ELECTION_TIMEOUT * 0.90
+    )  # sleep until right before election timeout
+    terms_passed = (
+        raft_node._term + 5
+    )  # leader's term is higher than the follower's term
```
format

maksimdrachov commented on tests/raft_node.py:
```diff
@@ -222,8 +314,12 @@ async def _unittest_raft_node_heartbeat() -> None:
 
     # send heartbeat again
     # (this time from a different leader with a higher term, we want to make sure the follower switches leader and updates term)
-    await asyncio.sleep(ELECTION_TIMEOUT * 0.90)  # sleep until right before election timeout
-    terms_passed = raft_node._term + 5  # leader's term is higher than the follower's term
+    await asyncio.sleep(
+        ELECTION_TIMEOUT * 0.90
+    )  # sleep until right before election timeout
+    terms_passed = (
+        raft_node._term + 5
+    )  # leader's term is higher than the follower's term
```
format

Parsifal22 replied:
[Fixed](https://github.com/OpenCyphal-Garage/cyraft/pull/11/commits/41b165dea3c17409abcae2e1fb337c45846b9747)

maksimdrachov commented on tests/raft_node.py:
```diff
@@ -248,8 +344,12 @@ async def _unittest_raft_node_heartbeat() -> None:
 
     # send heartbeat again
     # (this time the leader's term is lower than the follower's term, we want to make sure the follower doesn't switch leader)
-    await asyncio.sleep(ELECTION_TIMEOUT * 0.90)  # sleep until right before election timeout
-    terms_passed = raft_node._term - 1  # leader's term is lower than the follower's term
+    await asyncio.sleep(
+        ELECTION_TIMEOUT * 0.90
+    )  # sleep until right before election timeout
+    terms_passed = (
+        raft_node._term - 1
+    )  # leader's term is lower than the follower's term
```
format

Parsifal22 replied:
[Fixed](https://github.com/OpenCyphal-Garage/cyraft/pull/11/commits/41b165dea3c17409abcae2e1fb337c45846b9747)

maksimdrachov commented on tests/raft_node.py:
```diff
@@ -271,71 +371,11 @@ async def _unittest_raft_node_heartbeat() -> None:
     assert raft_node._prev_state == RaftState.CANDIDATE
 
     raft_node.close()
-    await asyncio.sleep(1)  # fixes when just running this test, however not when "pytest /cyraft" is run
+    await asyncio.sleep(
+        1
+    )  # fixes when just running this test, however not when "pytest /cyraft" is run
```
format

Parsifal22 replied:
[Fixed](https://github.com/OpenCyphal-Garage/cyraft/pull/11/commits/41b165dea3c17409abcae2e1fb337c45846b9747)

maksimdrachov commented on tests/raft_node.py:
```diff
@@ -271,71 +371,11 @@ async def _unittest_raft_node_heartbeat() -> None:
-        transfer_id=0,
-    )
-
-    response = await raft_node._serve_request_vote(request, metadata)
-    assert raft_node._voted_for == 43
-    assert response.vote_granted == False
-
-    # test 2: vote not granted if follower node term is greater than candidate's term
-    raft_node._voted_for = None  # node has not voted for any candidate
-    raft_node._term = request.term + 1
-
-    assert request.term < raft_node._term  # follower node term is greater than candidate's term
-    response = await raft_node._serve_request_vote(request, metadata)
-    assert raft_node._voted_for == None
-    assert response.vote_granted == False
-
-    # test 3: vote granted if not voted for another candidate
-    #         and the candidate's term is GREATER than the node's term
-    raft_node._voted_for = None
-    raft_node._term = request.term - 1
-
-    assert request.term > raft_node._term  # follower node term is less than candidate's term
-    response = await raft_node._serve_request_vote(request, metadata)
-    assert raft_node._voted_for == 42
-    assert response.vote_granted == True
-    assert raft_node._term == request.term  # follower node term is updated to candidate's term
-
-    # test 4: vote granted if not voted for another candidate
-    #         and the candidate's term is EQUAL to the node's term
-    raft_node._voted_for = None
-    raft_node._term = request.term - 1
-
-    assert request.term > raft_node._term  # follower node term is less than candidate's term
-    response = await raft_node._serve_request_vote(request, metadata)
-    assert raft_node._voted_for == 42
-    assert response.vote_granted == True
-    assert raft_node._term == request.term  # follower node term is updated to candidate's term
-    raft_node.close()
-    await asyncio.sleep(1)
-
```
Why did you remove this test?

Parsifal22 replied:
In the last commit all 13 tests are in place

maksimdrachov replied:
Ok, but moving there is no reason to move the test.

Parsifal22 replied:
In the last [commit](https://github.com/OpenCyphal-Garage/cyraft/pull/11/commits/aceea5df5436e0b99ccaf7df931cecab0121198e) I did as it was in main

maksimdrachov commented on tests/raft_node.py:
```diff
@@ -699,7 +751,9 @@ async def _unittest_raft_node_append_entries_rpc() -> None:
-    _logger.info("================== TEST 6: Try to append old log entry (term < currentTerm) ==================")
+    _logger.info(
+        "================== TEST 6: Try to append old log entry (term < currentTerm) =================="
+    )
```
format

Parsifal22 replied:
[Fixed](https://github.com/OpenCyphal-Garage/cyraft/pull/11/commits/41b165dea3c17409abcae2e1fb337c45846b9747)

maksimdrachov commented:
Before requesting a review:

- make sure the formatting is correct (everywhere)
- make sure that you haven't removed any unit tests
- make sure that all changes that are made are relevant to this merge request's topic

Parsifal22 commented on tests/raft_leader_election.py:
```diff
@@ -149,9 +151,7 @@ async def _unittest_raft_fsm_1() -> None:
-    assert (
-        raft_node_1._term == 1
-    ), "+1 due to starting election"
+    assert raft_node_1._term == 1, "+1 due to starting election"
```
[Fixed](https://github.com/OpenCyphal-Garage/cyraft/pull/11/commits/41b165dea3c17409abcae2e1fb337c45846b9747)

Parsifal22 commented on tests/raft_leader_election.py:
```diff
@@ -136,7 +136,9 @@ async def _unittest_raft_fsm_1() -> None:
-    _logger.info("================== TEST STAGE 2/3: LEADER election ==================")
+    _logger.info(
+        "================== TEST STAGE 2/3: LEADER election =================="
+    )
```
[Fixed](https://github.com/OpenCyphal-Garage/cyraft/pull/11/commits/41b165dea3c17409abcae2e1fb337c45846b9747)

Parsifal22 commented on tests/raft_leader_election.py:
```diff
@@ -164,10 +164,16 @@ async def _unittest_raft_fsm_1() -> None:
-    _logger.info("================== TEST STAGE 4: LEADER maintains leadership ==================")
+    _logger.info(
+        "================== TEST STAGE 4: LEADER maintains leadership =================="
+    )
```
[Fixed](https://github.com/OpenCyphal-Garage/cyraft/pull/11/commits/41b165dea3c17409abcae2e1fb337c45846b9747)

Parsifal22 commented on tests/raft_leader_election.py:
```diff
@@ -192,7 +198,9 @@ async def _unittest_raft_fsm_1() -> None:
-    _logger.info("================== DOES SOMETHING HAPPEN AFTER THIS? ==================")
+    _logger.info(
+        "================== DOES SOMETHING HAPPEN AFTER THIS? =================="
+    )
```
[Fixed](https://github.com/OpenCyphal-Garage/cyraft/pull/11/commits/41b165dea3c17409abcae2e1fb337c45846b9747)

Parsifal22 commented on tests/raft_leader_election.py:
```diff
@@ -207,7 +215,9 @@ async def _unittest_raft_fsm_2():
-    _logger.info("================== TEST STAGE 5/6: node with higher term gets elected ==================")
+    _logger.info(
+        "================== TEST STAGE 5/6: node with higher term gets elected =================="
+    )
```
[Fixed](https://github.com/OpenCyphal-Garage/cyraft/pull/11/commits/41b165dea3c17409abcae2e1fb337c45846b9747)
